### PR TITLE
Add two new OGC functions ST_X and ST_Y

### LINKED ISSUE #107325: Support initial geometry access commands ST_X and ST_Y
Recently Nik did work that involved extracting the X and Y coordinates from geo_point data using `to_string(field)` followed by a DISSECT command to re-parse the string to get the X and Y coordinates.

This is much more efficiently achieved using existing known OGC functions `ST_X` and `ST_Y`.

Tasks:

- [x] Initial prototype
- [x] Add unit tests and check that types files are generated correctly
- [x] Add docs for these functions

### PR DESCRIPTION
**Pull request by craigtaverner**
Recently Nik did work that involved extracting the X and Y coordinates from geo_point data using `to_string(field)` followed by a DISSECT command to re-parse the string to get the X and Y coordinates.

This is much more efficiently achieved using existing known OGC functions `ST_X` and `ST_Y`.

Still to do:

- [x] Add unit tests and check that types files are generated correctly
- [x] Add docs for these functions

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StX.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import org.elasticsearch.xpack.ql.type.DataType;
+
+import java.util.List;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.esql.expression.EsqlTypeResolutions.isSpatialPoint;
+import static org.elasticsearch.xpack.ql.type.DataTypes.DOUBLE;
+import static org.elasticsearch.xpack.ql.util.SpatialCoordinateTypes.UNSPECIFIED;
+
+public class StX extends UnaryScalarFunction {
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StX.java:29]: Probably worth javadoc linking to the spec to the function and mentioning how it works in geo and cartesian. I know it's obvious to geo folks but not so much for me.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StX.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    @FunctionInfo(returnType = "double", description = "Extracts the x-coordinate from a point geometry.")
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StX.java:30]: might want to say that it's lat or lon from geo.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java [hunk 3]
```diff
@@ -56,16 +56,30 @@ public long pointAsLong(double x, double y) {
             final long yi = XYEncodingUtils.encode((float) y);
             return (yi & 0xFFFFFFFFL) | xi << 32;
         }
+    },
+    UNSPECIFIED {
+        public Point longAsPoint(long encoded) {
```
Comment (nik9000) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:61]: It'd be nice if we could pass a `Point` as a spare and write the results to it. Or something like that. Like we do with `BytesRef`.

**Review by nik9000 — APPROVED** [review]
LGTM. @iverase have an opinion on this one?

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java (hunk 3 above)
Comment (craigtaverner) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:61]: I think I could add an extra method to WellKnownBinary class to extract this information without ever creating an intermediate object.

Comment (iverase) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:61]: I don't know, I think WKB should always be treated as a geo_shape, need to understand why we are trying to cast wkb to a long, feels wrong.

Comment (iverase) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:61]: and I think this `UNSPECIFIED` should go away, it makes no sense to me

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StX.java (hunk 1 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StX.java:36]: Added links

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java (hunk 3 above)
Comment (craigtaverner) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:61]: Added direct access from the WKB, but there is still one object creation for the big-ending/little-endian control.

Comment (iverase) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:61]: sorry but -1 to hack wkb to access values directly

Comment (iverase) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/SpatialCoordinateTypes.java:61]: big -1

File: libs/geo/src/main/java/org/elasticsearch/geometry/utils/WellKnownBinary.java [hunk 4]
```diff
@@ -217,6 +217,19 @@ public static Geometry fromWKB(GeometryValidator validator, boolean coerce, byte
         return geometry;
     }
 
+    /**
+     * Reads a point coordinate from the given WKB byte array with offset.
+     */
+    public static double coordinateFromWKB(int coordinateIndex, byte[] wkb, int offset, int length) {
```
Comment (iverase) [inline libs/geo/src/main/java/org/elasticsearch/geometry/utils/WellKnownBinary.java:223]: -1 to add such a method here sorry

**Review by iverase — CHANGES_REQUESTED** [review]
I am ok as far as we remove the new method in WellKnownBinary. We should not add such a non-standard stuff on that library.

File: libs/geo/src/main/java/org/elasticsearch/geometry/utils/WellKnownBinary.java (hunk 4 above)
Comment (iverase) [reply libs/geo/src/main/java/org/elasticsearch/geometry/utils/WellKnownBinary.java:223]: I think you can just move it as a private method where you need it.

Comment (craigtaverner) [reply libs/geo/src/main/java/org/elasticsearch/geometry/utils/WellKnownBinary.java:223]: OK. I'll remove that. I added based on Nik's comment, but he did not strictly need it, just an idea to reduce object creation.

Comment (iverase) [reply libs/geo/src/main/java/org/elasticsearch/geometry/utils/WellKnownBinary.java:223]: You can add it but not in the library. Let's keep it simple this library, read from / write to wkb. If you want to add that optimisation, add it in the ESQL module where you need it,

**Review by iverase — APPROVED** [review]

File: docs/reference/esql/functions/st_x.asciidoc [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+image::esql/functions/signature/st_x.svg[Embedded,opts=inline]
+
+*Parameters*
+
+`point`::
+Expression of type `geo_point` or `cartesian_point`. If `null`, the function returns `null`.
+
+*Description*
+
+Special function extracts the `x` coordinate from the supplied point.
```
Comment (astefan) [inline docs/reference/esql/functions/st_x.asciidoc:17]: I think the description should be ok without the "Special function" wording. To date I don't think we have any special functions.

File: docs/reference/esql/functions/st_y.asciidoc [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+image::esql/functions/signature/st_y.svg[Embedded,opts=inline]
+
+*Parameters*
+
+`point`::
+Expression of type `geo_point` or `cartesian_point`. If `null`, the function returns `null`.
+
+*Description*
+
+Special function extracts the `x` coordinate from the supplied point.
```
Comment (astefan) [inline docs/reference/esql/functions/st_y.asciidoc:17]: Same here about the "special" wording.

Comment (astefan) [reply docs/reference/esql/functions/st_y.asciidoc:17]: Also, I think this is copy-paste and you forgot to adapt it to `st_y` (for example, in this line you say "extracts the `x` coordinate).

File: docs/reference/esql/functions/st_y.asciidoc [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+If the points is of type `geo_point` this is equivalent to extracting the `longitude` value.
```
Comment (astefan) [inline docs/reference/esql/functions/st_y.asciidoc:18]: Same for `longitude`. If I'm not mistaken, it should be latitude.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StY.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import static org.elasticsearch.xpack.ql.util.SpatialCoordinateTypes.UNSPECIFIED;
+
+/**
+ * Extracts the y-coordinate from a point geometry.
+ * For cartesian geometries, the y-coordinate is the second coordinate.
+ * For geographic geometries, the y-coordinate is the latitude.
+ * The function `st_y` is defined in the <a href="https://www.ogc.org/standard/sfs/">OGC Simple Feature Access</a> standard.
+ * Alternatively it is well described in PostGIS documentation at <a href="https://postgis.net/docs/ST_Y.html">PostGIS:ST_Y</a>.
+ */
+public class StY extends UnaryScalarFunction {
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StY.java:36]: StX and StY are very similar. You could create a common class for them with (from the first look) `resolveType` and `dataType` as common methods.

File: docs/reference/esql/functions/st_x.asciidoc (hunk 5 above)
Comment (craigtaverner) [reply docs/reference/esql/functions/st_x.asciidoc]: Indeed, there is nothing particularly 'special' about these functions!
Fixed in 379d6cdd5577a5def64ae4fece1351bac7e64da3

File: docs/reference/esql/functions/st_y.asciidoc (hunk 6 above)
Comment (craigtaverner) [reply docs/reference/esql/functions/st_y.asciidoc]: I thought I double-checked this, but clearly not! Fixed in 379d6cdd5577a5def64ae4fece1351bac7e64da3

File: docs/reference/esql/functions/st_y.asciidoc (hunk 7 above)
Comment (craigtaverner) [reply docs/reference/esql/functions/st_y.asciidoc]: Fixed in 379d6cdd5577a5def64ae4fece1351bac7e64da3

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StY.java (hunk 8 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StY.java:36]: I thought about that, but the amount of duplicated code is so tiny, and the classes themselves are so tiny, that this seemed to increase complexity more than it reduced it. I thought instead to see once we had a lot more `ST_*` functions to decide then if there was reasonable commonality to create parent classes.

**Pull request merged**